File: packages/js/core/src/workflow/Workflow.ts
import { ClientConfig, Uri } from "..";
import { InvokeResult } from "../invocation/Invoke";
import { MaybeAsync } from "../../utils/MaybeAsync";

export type Step<TUri extends Uri | string = string> = {
  uri: TUri;
  method: string;
  args: Record<string, unknown>;
  config?: Partial<ClientConfig>;
};

export type JobInfo<TUri extends Uri | string = string> = {
  steps?: Step<TUri>[];
  jobs?: Job<TUri>;
};

export type Job<TUri extends Uri | string = string> = {
  [name: string]: JobInfo<TUri>;
};

export type Workflow<TUri extends Uri | string = string> = {
  name: string;
  jobs: Job<TUri>;
};

File: packages/py/core/core/types/workflow.py
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING, Any, Dict, List, Optional

from .invoke import InvokeResult
from .uri import Uri

if TYPE_CHECKING:
    from .client import ClientConfig


@dataclass(slots=True, kw_only=True)
class Step:
    uri: Uri
    method: str
    args: Dict[str, Any] = field(default_factory=dict)
    config: Optional["ClientConfig"] = None


Job = Dict[str, "JobInfo"]


@dataclass(slots=True, kw_only=True)
class JobInfo:
    steps: List[Step]
    jobs: List[Job]


@dataclass(slots=True, kw_only=True)
class Workflow:
    name: str
    jobs: Job


class JobStatus(Enum):
    SUCCEED = "SUCCEED"
    FAILED = "FAILED"
    SKIPPED = "SKIPPED"


@dataclass(slots=True, kw_only=True)
class JobResult(InvokeResult):
    status: JobStatus


@dataclass(slots=True, kw_only=True)
class RunOptions:
    workflow: Workflow
    config: Optional["ClientConfig"] = None
    context_id: Optional[str] = None
    ids: Optional[List[str]] = field(default_factory=list)


class WorkflowHandler(ABC):
    @abstractmethod
    async def run(self, options: RunOptions) -> None:
        pass
